File: java/leetcode/src/main/java/com/leetcode/tree/Symmetric_Tree_101.java
package com.leetcode.tree;

public class Symmetric_Tree_101 {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;
        return isSymmetric(root.left, root.right);
    }

    boolean isSymmetric(TreeNode p, TreeNode q) {
        if (q == null && p == null) return true;
        if ((p == null && q != null) || (p != null && q == null)) return false;
        if (p.val != q.val) return false;
        return isSymmetric(p.left, q.right) && isSymmetric(p.right, q.left);
    }
}

File: js/algrithm/huahua130/tree/Symmetric_Tree_101.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    const isMirror = function(p, q) {
        if (!p && !q) return true
        if ((p && !q) || (!p && q)) return false
        if (p && q && p.val !== q.val) return false
        const outer = isMirror(p.left, q.right)
        const iner = isMirror(p.right, q.left)
        return outer && iner
    }
    const res = isMirror(root.left, root.right)
    return res
};